packages/dd-trace/src/debugger/devtools_client/condition.js

```diff
@@ -0,0 +1,237 @@
+            return String.prototype.substring.call(str, ${args[1]}, ${args[2]})
+          } else {
+            throw new TypeError('Variable is not a string')
+          }
+        })(${args[0]})`
+      case 'getmember': return accessProperty(args[0], args[1], false)
+      case 'index': return accessProperty(args[0], args[1], true)
+    }
+  }
+
+  throw new TypeError(`Unknown AST node type: ${type}`)
+}
+
+function iterateOn (fnName, variable, callbackCode) {
+  return `(($dd_val) => {
+    return ${isIterableCollection('$dd_val')}
+      ? Array.from($dd_val).${fnName}(($dd_it) => ${callbackCode})
+      : Object.entries($dd_val).${fnName}(([$dd_key, $dd_value]) => ${callbackCode})
+  })(${variable})`
+}
+
+function isString (variable) {
+  return `typeof ${variable} === 'string' || ${variable} instanceof String`
+}
+
+function isIterableCollection (variable) {
+  return `${isArrayOrTypedArray(variable)} || ${variable} instanceof Set || ${variable} instanceof WeakSet`
+}
+
+function isArrayOrTypedArray (variable) {
+  return `Array.isArray(${variable}) || ${variable} instanceof Object.getPrototypeOf(Int8Array)`
+}
+
+function getSize (variable) {
+  return `((val) => {
+    if (${isString('val')} || ${isArrayOrTypedArray('val')}) {
+      return ${guardAgainstPropertyAccessSideEffects('val', '"length"')}
+    } else if (val instanceof Set || val instanceof Map) {
+      return ${guardAgainstPropertyAccessSideEffects('val', '"size"')}
+    } else {
+      throw new TypeError('Cannot get length or size of string/collection')
+    }
+  })(${variable})`
+}
+
+function accessProperty (variable, keyOrIndex, allowMapAccess) {
+  return `((val, key) => {
+    if (val instanceof Set || val instanceof WeakSet) {
+      throw new Error('Accessing a Set or WeakSet is not allowed')
+    } else if (val instanceof Map) {
+      ${allowMapAccess
+        ? 'return Map.prototype.get.call(val, key)'
+        : 'throw new Error(\'Accessing a Map is not allowed\')'}
+    } else if (val instanceof WeakMap) {
+      ${allowMapAccess
+        ? 'return WeakMap.prototype.get.call(val, key)'
+        : 'throw new Error(\'Accessing a WeakMap is not allowed\')'}
+    } else {
+      return ${guardAgainstPropertyAccessSideEffects('val', 'key')}
+    }
+  })(${variable}, ${keyOrIndex})`
+}
+
+function guardAgainstPropertyAccessSideEffects (variable, propertyName) {
+  return `((val, key) => {
+    if (${maybeProxy('val')} || Object.getOwnPropertyDescriptor(val, key)?.get !== undefined) {
+      throw new Error('Possibility of side effect')
+    } else {
+      return val[key]
+    }
+  })(${variable}, ${propertyName})`
+}
+
+function guardAgainstCoercionSideEffects (variable) {
+  return `((val) => {
+    if (
+      typeof val === 'object' && val !== null && (
+        ${maybeProxy('val')} ||
+        val[Symbol.toPrimitive] !== undefined ||
+        val.valueOf !== Object.prototype.valueOf ||
+        val.toString !== Object.prototype.toString
```

Comment:
Are arrays allowed? These have another `.toString()` property.

Reply:
No I don't think there would be any reason to compare an array using `<` or `>`.